src/theory/sets/theory_sets_type_enumerator.h

```diff
@@ -2,185 +2,164 @@
+{
  public:
   SetEnumerator(TypeNode type, TypeEnumeratorProperties* tep = nullptr)
       : TypeEnumeratorBase<SetEnumerator>(type),
-        d_tep(tep),
-        d_index(0),
-        d_constituentType(type.getSetElementType()),
-        d_nm(NodeManager::currentNM()),
-        d_indexVec(),
-        d_constituentVec(),
-        d_finished(false),
-        d_setConst()
+        d_nodeManager(NodeManager::currentNM()),
+        d_elementEnumerator(type.getSetElementType(), tep),
+        d_isFinished(false),
+        d_currentSetIndex(0),
+        d_currentSet()
   {
-    // d_indexVec.push_back(false);
-    // d_constituentVec.push_back(new TypeEnumerator(d_constituentType));
-    d_setConst = d_nm->mkConst(EmptySet(type.toType()));
+    d_currentSet = d_nodeManager->mkConst(EmptySet(type.toType()));
   }
 
-  // An set enumerator could be large, and generally you don't want to
-  // go around copying these things; but a copy ctor is presently required
-  // by the TypeEnumerator framework.
-  SetEnumerator(const SetEnumerator& ae)
-      : TypeEnumeratorBase<SetEnumerator>(
-            ae.d_nm->mkSetType(ae.d_constituentType)),
-        d_tep(ae.d_tep),
-        d_index(ae.d_index),
-        d_constituentType(ae.d_constituentType),
-        d_nm(ae.d_nm),
-        d_indexVec(ae.d_indexVec),
-        d_constituentVec(),  // copied below
-        d_finished(ae.d_finished),
-        d_setConst(ae.d_setConst)
+  SetEnumerator(const SetEnumerator& enumerator)
+      : TypeEnumeratorBase<SetEnumerator>(enumerator.getType()),
+        d_nodeManager(enumerator.d_nodeManager),
+        d_elementEnumerator(enumerator.d_elementEnumerator),
+        d_isFinished(enumerator.d_isFinished),
+        d_currentSetIndex(enumerator.d_currentSetIndex),
+        d_currentSet(enumerator.d_currentSet)
   {
-    for(std::vector<TypeEnumerator*>::const_iterator i =
-          ae.d_constituentVec.begin(), i_end = ae.d_constituentVec.end();
-        i != i_end;
-        ++i) {
-      d_constituentVec.push_back(new TypeEnumerator(**i));
-    }
   }
 
-  ~SetEnumerator() {
-    while (!d_constituentVec.empty()) {
-      delete d_constituentVec.back();
-      d_constituentVec.pop_back();
-    }
-  }
+  ~SetEnumerator() {}
 
   Node operator*() override
   {
-    if (d_finished) {
+    if (d_isFinished)
+    {
       throw NoMoreValuesException(getType());
     }
 
-    std::vector<Node> elements;
-    for(unsigned i = 0; i < d_constituentVec.size(); ++i) {
-      elements.push_back(*(*(d_constituentVec[i])));
-    }
-
-    Node n = NormalForm::elementsToSet(std::set<TNode>(elements.begin(), elements.end()),
-                                       getType());
+    Trace("set-type-enum") << "SetEnumerator::operator* d_currentSet = "
+                           << d_currentSet << std::endl;
 
-    Assert(n.isConst());
-    Assert(n == Rewriter::rewrite(n));
-
-    return n;
+    return d_currentSet;
   }
 
+  /**
```

Comment:
Can you move this (and other functions) to a *.cpp file?

Reply:
done